### PR TITLE
Core/Creatures: Implemented UNIT_NPC_FLAG_GOSSIP as viewer dependent …

### PR DESCRIPTION
**Pull request by mdX7**
…and update npcflags on quest changes

**Tests performed:**
tested ingame, utilizing gossip menu conditions

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -27098,6 +27138,43 @@ bool Player::CanSeeSpellClickOn(Creature const* c) const
...
+        {
+            if (GetGossipMenuForSource(creature))
+                return true;
+        }
+    }
+
+    // for cases with questgiver/ender without gossip menus
+    if (creature->HasNpcFlag(UNIT_NPC_FLAG_QUESTGIVER))
+    {
+        QuestRelationResult objectQR = sObjectMgr->GetCreatureQuestRelations(creature->GetEntry());
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp:27155]: Move this to just before loop (so that its never called if any of objectQIR returns true)

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by killerwife** [comment]
What expansion is this used in

**Comment by mdX7** [comment]
I've seen it minimum in BfA

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -2032,6 +2032,34 @@ GameObject* Player::GetGameObjectIfCanInteractWith(ObjectGuid const& guid, Gameo
...
+        // skip creatures which dont have any npcflags set
+        if (!creature->GetNpcFlags() && !creature->GetNpcFlags2())
+            continue;
+
+        creature->BuildValuesUpdateForPlayerWithMask(&udata, objMask.GetChangesMask(), unitMask.GetChangesMask(), this);
+    }
+
+    WorldPacket packet;
+    udata.BuildPacket(&packet);
+    SendDirectMessage(&packet);
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp:2060]:
```suggestion
    if (!udata.HasData())
        return;

    WorldPacket packet;
    udata.BuildPacket(&packet);
    SendDirectMessage(&packet);
```

**Review by Shauren — APPROVED** [review]

**Pull request merged**